### ISSUE TITLE
buildFHSEnv environments get garbage collected while open

### ISSUE THREAD

**Issue by enolan** [NixOS/nixpkgs#365399]
I have a Python shell environment created with `buildFHSEnv` using a flake. If gc runs while I have the environment open with `nix run`, python disappears. Presumably also anything else used in the environment that otherwise doesn't have a gc root.

## Steps To Reproduce

1. Put the `flake.nix` and `flake.lock` from [this gist](https://gist.github.com/enolan/9fe4b5a4b70a850613c4ac5de2d01f9f) in a directory
2. `nix run $DIR`
3. run `python` in the shell that makes, notice it works.
4. in a separate terminal, run `nix-store --gc`
5. run `python` again in the `nix run` shell, notice the executable is gone

## Expected behavior

GC shouldn't delete things that are reachable from virtual environments like `buildFHSEnv`, for the same reason it shouldn't delete currently open executables or libraries. If I use `mkShell` and `nix develop` instead of `buildFHSEnv` and `nix run` it behaves as expected.

## Notify maintainers

Don't know who to ping here. Also not sure if this should be filed in `nixpkgs` or `nix` itself.

---

Note for maintainers: Please tag this issue in your PR.

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

**Comment by zimward** [comment]
I don't think that this is specifically linked to `buildFHSEnv` as it only builds a package like everything else in the devenv. It's more likely a edgecase of `nix run` which should likely setup gc-roots until the shell exits.
